File: practice-cpp/fundamentals/src/arreglos.cpp
#include <iostream>
#include <array>
#include <vector>
#include <algorithm>

using namespace std;

int main() {

    // ===== ARRAYS C-STYLE =====

    cout << "\n===== ARRAYS C-STYLE =====\n";

    int arraySimple[5] = {10, 20, 30, 40, 50};
    cout << "Array: ";
    for (int i = 0; i < 5; i++) {
        cout << arraySimple[i] << " ";
    }
    cout << "\n";

    // Acceso a elementos
    cout << "Primer elemento: " << arraySimple[0] << "\n";
    cout << "Último elemento: " << arraySimple[4] << "\n";
    cout << "Tamaño: 5\n";

    // Modificar un elemento
    arraySimple[0] = 100;
    cout << "Después de modificar: " << arraySimple[0] << "\n";

    // ===== STD::ARRAY (C++11) =====

    cout << "\n===== STD::ARRAY (C++11) =====\n";

    array<int, 4> arrayEstandar = {15, 25, 35, 45};
    cout << "std::array: ";
    for (int num : arrayEstandar) {
        cout << num << " ";
    }
    cout << "\n";
    cout << "Tamaño con .size(): " << arrayEstandar.size() << "\n";

    // ===== STD::VECTOR =====

    cout << "\n===== STD::VECTOR =====\n";

    vector<int> vector1;
    vector1.push_back(100);
    vector1.push_back(200);
    vector1.push_back(300);
    vector1.push_back(400);

    cout << "Vector: ";
    for (int val : vector1) {
        cout << val << " ";
    }
    cout << "\n";
    cout << "Tamaño: " << vector1.size() << "\n";
    cout << "Primer elemento: " << vector1[0] << "\n";
    cout << "Último elemento: " << vector1.back() << "\n";

    // Agregar más elementos
    vector1.push_back(500);
    cout << "Después de push_back(500): ";
    for (int val : vector1) {
        cout << val << " ";
    }
    cout << "\n";

    // Acceso seguro con .at()
    cout << "Acceso con .at(2): " << vector1.at(2) << "\n";

    // Eliminar último elemento
    vector1.pop_back();
    cout << "Después de pop_back(): tamaño = " << vector1.size() << "\n";

    // ===== RECORRIDO CON FOR CLÁSICO =====

    cout << "\n===== RECORRIDO CON FOR CLÁSICO =====\n";

    int calificaciones[] = {85, 92, 78, 88, 95};
    cout << "Calificaciones:\n";
    for (int i = 0; i < 5; i++) {
        cout << "  Posición " << i << ": " << calificaciones[i] << "\n";
    }

    // ===== RECORRIDO CON RANGE-BASED FOR =====

    cout << "\n===== RECORRIDO CON RANGE-BASED FOR (C++11) =====\n";

    string estudiantes[] = {"Ana", "Bruno", "Carlos", "Diana"};
    cout << "Lista de estudiantes:\n";
    for (const string& estudiante : estudiantes) {
        cout << "  - " << estudiante << "\n";
    }

    // ===== BÚSQUEDA EN VECTOR =====

    cout << "\n===== BÚSQUEDA EN VECTOR =====\n";

    vector<int> numeros2 = {10, 25, 30, 45, 50, 65};
    int buscar = 45;
    bool encontrado = false;

    for (int numero : numeros2) {
        if (numero == buscar) {
            encontrado = true;
            break;
        }
    }

    cout << "¿" << buscar << " está en el vector? " << (encontrado ? "Sí" : "No") << "\n";

    // ===== SORT: ORDENAR VECTOR =====

    cout << "\n===== SORT: ORDENAR VECTOR =====\n";

    vector<int> numerosDesordenados = {45, 12, 78, 23, 56, 89, 34};
    cout << "Antes de ordenar: ";
    for (int n : numerosDesordenados) cout << n << " ";
    cout << "\n";

    sort(numerosDesordenados.begin(), numerosDesordenados.end());
    cout << "Después de ordenar: ";
    for (int n : numerosDesordenados) cout << n << " ";
    cout << "\n";

    // Ordenar strings
    vector<string> frutas = {"Manzana", "Cereza", "Banana", "Durazno"};
    cout << "Frutas desordenadas: ";
    for (const string& f : frutas) cout << f << " ";
    cout << "\n";
    sort(frutas.begin(), frutas.end());
    cout << "Frutas ordenadas: ";
    for (const string& f : frutas) cout << f << " ";
    cout << "\n";

    // ===== SUMA Y PROMEDIO =====

    cout << "\n===== SUMA Y PROMEDIO =====\n";

    vector<int> precios = {100, 250, 150, 300, 200};
    int suma = 0;

    for (int precio : precios) {
        suma += precio;
    }

    double promedio = static_cast<double>(suma) / precios.size();
    cout << "Array de precios: ";
    for (int p : precios) cout << p << " ";
    cout << "\n";
    cout << "Suma total: " << suma << "\n";
    cout << "Promedio: " << promedio << "\n";

    // ===== MÁXIMO Y MÍNIMO =====

    cout << "\n===== MÁXIMO Y MÍNIMO =====\n";

    vector<int> temperaturas = {15, 22, 28, 18, 25, 30, 20};
    int maxima = *max_element(temperaturas.begin(), temperaturas.end());
    int minima = *min_element(temperaturas.begin(), temperaturas.end());

    cout << "Temperaturas: ";
    for (int t : temperaturas) cout << t << " ";
    cout << "\n";
    cout << "Temperatura máxima: " << maxima << "°C\n";
    cout << "Temperatura mínima: " << minima << "°C\n";

    // ===== ARRAYS MULTIDIMENSIONALES (2D - MATRIZ) =====

    cout << "\n===== ARRAYS MULTIDIMENSIONALES (MATRIZ 2D) =====\n";

    // matriz de 3 filas y 3 columnas
    int matriz[3][3] = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };

    cout << "Matriz 3x3:\n";
    for (int fila = 0; fila < 3; fila++) {
        for (int columna = 0; columna < 3; columna++) {
            cout << matriz[fila][columna] << " ";
        }
        cout << "\n";
    }

    cout << "Elemento en posición [1][2]: " << matriz[1][2] << "\n";

    // ===== VECTOR DE VECTORES (MATRIZ DINÁMICA) =====

    cout << "\n===== VECTOR DE VECTORES (MATRIZ DINÁMICA) =====\n";

    vector<vector<int>> matrizDinamica = {
        {10, 20, 30},
        {40, 50, 60},
        {70, 80, 90}
    };

    cout << "Matriz dinámica:\n";
    for (const auto& fila : matrizDinamica) {
        for (int elem : fila) {
            cout << elem << " ";
        }
        cout << "\n";
    }

    return 0;
}

File: practice-java/fundamentals/src/Arreglos.java
public class Arreglos {
    public static void main(String[] args) {

        // ===== DECLARACIÓN E INICIALIZACIÓN DE ARRAYS =====

        System.out.println("===== DECLARACIÓN E INICIALIZACIÓN =====");

        // forma 1: declarar con tamaño fijo, elementos iniciados a 0
        int[] numeros = new int[5];
        System.out.println("Array de 5 enteros: " + java.util.Arrays.toString(numeros));

        // forma 2: declarar e inicializar con valores específicos
        String[] colores = {"Rojo", "Verde", "Azul", "Amarillo"};
        System.out.println("Array de colores: " + java.util.Arrays.toString(colores));

        // forma 3: declarar con valores y el compilador deduce el tamaño
        int[] pares = {2, 4, 6, 8, 10};
        System.out.println("Array de pares: " + java.util.Arrays.toString(pares));

        // ===== ACCESO A ELEMENTOS POR ÍNDICE =====

        System.out.println("\n===== ACCESO POR ÍNDICE =====");

        // los índices comienzan en 0
        String[] diasSemana = {"Lunes", "Martes", "Miércoles", "Jueves", "Viernes", "Sábado", "Domingo"};

        System.out.println("Primer día (índice 0): " + diasSemana[0]);
        System.out.println("Tercer día (índice 2): " + diasSemana[2]);
        System.out.println("Último día (índice 6): " + diasSemana[6]);
        System.out.println("Total de días: " + diasSemana.length);

        // modificar un elemento
        diasSemana[0] = "Lunes (modificado)";
        System.out.println("Primer día después de modificar: " + diasSemana[0]);

        // ===== RECORRIDO CON FOR CLÁSICO =====

        System.out.println("\n===== RECORRIDO CON FOR CLÁSICO =====");

        int[] calificaciones = {85, 92, 78, 88, 95};
        System.out.println("Calificaciones:");
        for (int i = 0; i < calificaciones.length; i++) {
            System.out.println("  Posición " + i + ": " + calificaciones[i]);
        }

        // ===== RECORRIDO CON FOR-EACH =====

        System.out.println("\n===== RECORRIDO CON FOR-EACH =====");

        String[] estudiantes = {"Ana", "Bruno", "Carlos", "Diana"};
        System.out.println("Lista de estudiantes:");
        for (String estudiante : estudiantes) {
            System.out.println("  - " + estudiante);
        }

        // ===== BÚSQUEDA EN ARRAY =====

        System.out.println("\n===== BÚSQUEDA EN ARRAY =====");

        int[] numeros2 = {10, 25, 30, 45, 50, 65};
        int buscar = 45;
        boolean encontrado = false;

        for (int numero : numeros2) {
            if (numero == buscar) {
                encontrado = true;
                break;
            }
        }

        System.out.println("¿" + buscar + " está en el array? " + encontrado);

        // ===== ARRAYS.SORT(): ORDENAR ARRAY =====

        System.out.println("\n===== ARRAYS.SORT() - ORDENAR ARRAY =====");

        int[] numerosDesordenados = {45, 12, 78, 23, 56, 89, 34};
        System.out.println("Antes de ordenar: " + java.util.Arrays.toString(numerosDesordenados));

        java.util.Arrays.sort(numerosDesordenados);
        System.out.println("Después de ordenar: " + java.util.Arrays.toString(numerosDesordenados));

        // ordenar strings (orden alfabético)
        String[] frutas = {"Manzana", "Cereza", "Banana", "Durazno"};
        System.out.println("Frutas desordenadas: " + java.util.Arrays.toString(frutas));
        java.util.Arrays.sort(frutas);
        System.out.println("Frutas ordenadas: " + java.util.Arrays.toString(frutas));

        // ===== SUMA Y PROMEDIO DE ELEMENTOS =====

        System.out.println("\n===== SUMA Y PROMEDIO =====");

        int[] precios = {100, 250, 150, 300, 200};
        int suma = 0;

        for (int precio : precios) {
            suma += precio;
        }

        double promedio = (double) suma / precios.length;
        System.out.println("Array de precios: " + java.util.Arrays.toString(precios));
        System.out.println("Suma total: " + suma);
        System.out.println("Promedio: " + promedio);

        // ===== MÁXIMO Y MÍNIMO =====

        System.out.println("\n===== MÁXIMO Y MÍNIMO =====");

        int[] temperaturas = {15, 22, 28, 18, 25, 30, 20};
        int maxima = temperaturas[0];
        int minima = temperaturas[0];

        for (int temp : temperaturas) {
            if (temp > maxima) maxima = temp;
            if (temp < minima) minima = temp;
        }

        System.out.println("Temperaturas: " + java.util.Arrays.toString(temperaturas));
        System.out.println("Temperatura máxima: " + maxima + "°C");
        System.out.println("Temperatura mínima: " + minima + "°C");

        // ===== ARRAYS MULTIDIMENSIONALES (2D - MATRIZ) =====

        System.out.println("\n===== ARRAYS MULTIDIMENSIONALES (MATRIZ 2D) =====");

        // matriz de 3 filas y 3 columnas
        int[][] matriz = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 9}
        };

        System.out.println("Matriz 3x3:");
        for (int fila = 0; fila < matriz.length; fila++) {
            for (int columna = 0; columna < matriz[fila].length; columna++) {
                System.out.print(matriz[fila][columna] + " ");
            }
            System.out.println();
        }

        System.out.println("\nElemento en posición [1][2]: " + matriz[1][2]);

        // ===== MATRIZ CON FOR-EACH =====

        System.out.println("\n===== RECORRIDO DE MATRIZ CON FOR-EACH =====");

        int[][] calificacionesAlumnos = {
            {85, 90, 88},
            {92, 87, 95},
            {78, 82, 80}
        };

        System.out.println("Calificaciones por alumno:");
        for (int[] alumno : calificacionesAlumnos) {
            System.out.println("Alumno: " + java.util.Arrays.toString(alumno));
        }

        // ===== COPIAR ARRAYS =====

        System.out.println("\n===== COPIAR ARRAYS =====");

        int[] original = {10, 20, 30, 40};
        int[] copia = java.util.Arrays.copyOf(original, original.length);

        System.out.println("Array original: " + java.util.Arrays.toString(original));
        System.out.println("Array copiado: " + java.util.Arrays.toString(copia));

        // modificar la copia
        copia[0] = 999;
        System.out.println("Después de modificar la copia:");
        System.out.println("Original: " + java.util.Arrays.toString(original));
        System.out.println("Copia: " + java.util.Arrays.toString(copia));

    }
}
